Refactor/extract component

aloysb opened the pull request:
This pull request does not add any features to the repository but breaks down existing code into components.

The code has been re-arranged as follows:
- components/ contains the components of the application
- hooks/ for the custom hooks
- types/ contains the shared types

It also add a couple of testing dependencies: @testing-library/user-event and @testing-library/hooks.

This pull request also adds testing to all the code that has been refactored.

Commit:
build(Dependencies): Add testing dependencies: @testing-library/hooks and @testing-library/user-event

Both of these dependencies are required to test custom hooks and simulate user events. They are part of the @testing-library ecosystem and therefore not additional dependencies as per se.

Commit:
refactor(types): Extract types in its own types folder

Commit:
feat(CastMember): Extract the CastMember component

Commit:
feat(Loader): Extract the Loader component

Commit:
feat(Show): Extract the Show component

Commit:
feat(ShowList): Extract the ShowsList component

Commit:
feat(useSearchShow): Extract the useSearchShow custom hook

Commit:
test(tests): Add tests fixtures (i.e. data for testing purposes)

Commit:
refactor(App.tsx): Refactor the main component, App.tsx and add integration tests

Commit:
docs(README): Complete readme, add comments on the refactoring phase

Commit:
fix(types): Fix the types issues by improving the fixtures (tests) and the type definition of the shows

The type didn't not allow any extra properties.
However, in reality the API return a lot more properties, we just don't
use them.
For sake of being as close as possible to real world example, I've
extended the type to allow extra properties.

abitbetterthanyesterday commented:
The build was failing due to typing.

The definition of the IShow returned by the API is incomplete. In reality, the API returns a lot more properties.
In an attempt to be as close to real world as possible for testing, I've allowed extra properties on the IShow even though we do not use them in the application.

Commit:
build(dependencies(TS)): Update TypeScript to the latest versions

Commit:
fix(package-manager) Switch-to-npm-over-yarn-for-consistency-with-CI-pipeline

There were some versioning issues between package. For sake of consistency with the CI pipeline, I am switching the project to npm as a package manager
